libbpf-tools/solisten.bpf.c

```diff
@@ -0,0 +1,88 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/* Copyright (c) 2021 Hengqi Chen */
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_endian.h>
+#include <bpf/bpf_tracing.h>
+#include "solisten.h"
+
+#define MAX_ENTRIES	10240
+#define AF_INET		2
+#define AF_INET6	10
+
+const volatile pid_t target_pid = 0;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, __u32);
+	__type(value, struct event);
+} values SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(value_size, sizeof(__u32));
+} events SEC(".maps");
+
+static int probe_entry(struct socket *sock, int backlog)
+{
+	__u64 pid_tgid = bpf_get_current_pid_tgid();
+	__u32 pid = pid_tgid >> 32;
+	__u32 tid = (__u32)pid_tgid;
+	__u16 family, type;
+	struct sock *sk = BPF_CORE_READ(sock, sk);
+	struct inet_sock *inet = (struct inet_sock *)sk;
+	struct event event = {
+		.time = bpf_ktime_get_ns(),
+		.pid = pid,
+		.backlog = backlog,
+	};
+
+	if (target_pid && target_pid != pid)
+		return 0;
+
+	family = BPF_CORE_READ(sk, __sk_common.skc_family);
+	type = BPF_CORE_READ(sock, type);
+	event.proto = ((__u32)family << 16) | type;
+	event.port = bpf_ntohs(BPF_CORE_READ(inet, inet_sport));
+	if (family == AF_INET)
+		event.addr[0] = BPF_CORE_READ(sk, __sk_common.skc_rcv_saddr);
+	else if (family == AF_INET6)
+		bpf_probe_read_kernel(event.addr, sizeof(event.addr), sk->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
+	bpf_get_current_comm(event.task, sizeof(event.task));
+
+	bpf_map_update_elem(&values, &tid, &event, BPF_ANY);
+	return 0;
+};
+
+static int probe_exit(void *ctx, int ret)
+{
+	__u32 tid = bpf_get_current_pid_tgid();
+	struct event *eventp;
+
+	eventp = bpf_map_lookup_elem(&values, &tid);
+	if (!eventp)
+		return 0;
+
+	eventp->time = bpf_ktime_get_ns() - eventp->time;
+	eventp->ret = ret;
+	bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU, eventp, sizeof(*eventp));
+	bpf_map_delete_elem(&values, &tid);
+	return 0;
+}
+
+SEC("kprobe/inet_listen")
+int BPF_KPROBE(inet_listen_entry, struct socket *sock, int backlog)
+{
+	return probe_entry(sock, backlog);
+}
+
+SEC("kretprobe/inet_listen")
```

Comment:
Can these 2 probes be replaced with a single `fexit/inet_listen` probe? Would let us remove the `values` map.

Reply:
Nice idea. Add `fexit` and fallback to `kprobe/kretprobe` if not available.